File: diagnostics/rq_dimension_audit/scripts/reproduce_factor_selection.py
from pathlib import Path

import numpy as np
import pandas as pd


out_dir = Path("diagnostics/rq_dimension_audit/output")
panel_dir = out_dir / "panels"
r_selection = pd.read_csv(out_dir / "r_selection.csv")
r_bai_ng = pd.read_csv(out_dir / "r_bai_ng_criteria.csv")
r_aw = pd.read_csv(out_dir / "r_aw_criteria.csv")
r_bai_eigen = pd.read_csv(out_dir / "r_bai_ng_eigenvalues.csv")
r_aw_eigen = pd.read_csv(out_dir / "r_aw_eigenvalues.csv")

selection_rows = []
bai_rows = []
aw_rows = []
bai_eigen_rows = []
aw_eigen_rows = []

for panel_path in sorted(panel_dir.glob("*.csv.gz")):
    variant = panel_path.name.removesuffix(".csv.gz")
    panel = pd.read_csv(panel_path, parse_dates=["ref.date"])
    dates = panel.pop("ref.date")
    assert panel.shape[0] == 153
    assert dates.is_monotonic_increasing and not dates.duplicated().any()
    assert np.isfinite(panel.to_numpy()).all()

    samples = {
        "full": panel.to_numpy(dtype=float),
        "pre_covid": panel.loc[dates <= "2019-12-01"].to_numpy(dtype=float),
    }
    for sample_name, data in samples.items():
        differences = np.diff(data, axis=0)
        standard_deviations = differences.std(axis=0, ddof=1)
        assert np.isfinite(standard_deviations).all()
        assert (standard_deviations > 0).all()
        standardized = (differences - differences.mean(axis=0)) / standard_deviations

        _, singular_values, right_vectors = np.linalg.svd(
            standardized,
            full_matrices=False,
        )
        n_periods, n_series = standardized.shape
        total_sum_squares = np.square(singular_values).sum()
        residual_variances = (
            total_sum_squares - np.cumsum(np.square(singular_values[:20]))
        ) / (n_periods * n_series)
        r_values = np.arange(1, 21)
        dimension_penalty = (n_series + n_periods) / (n_series * n_periods)
        ic1 = np.log(residual_variances) + r_values * dimension_penalty * np.log(
            1 / dimension_penalty
        )
        ic2 = np.log(residual_variances) + r_values * dimension_penalty * np.log(
            min(n_series, n_periods)
        )
        ic3 = np.log(residual_variances) + r_values * np.log(
            min(n_series, n_periods)
        ) / min(n_series, n_periods)
        r_ic1 = int(np.argmin(ic1) + 1)
        r_ic2 = int(np.argmin(ic2) + 1)
        r_ic3 = int(np.argmin(ic3) + 1)

        factor_scores = standardized @ right_vectors[:r_ic2].T
        lag_rows = []
        for index in range(6, n_periods):
            lag_rows.append(
                np.concatenate(
                    [factor_scores[index - lag] for lag in range(1, 7)]
                )
            )
        design = np.column_stack([np.ones(n_periods - 6), np.vstack(lag_rows)])
        dependent = standardized[6:]
        coefficients = np.linalg.solve(design.T @ design, design.T @ dependent)
        residuals = dependent - design @ coefficients
        _, aw_singular_values, _ = np.linalg.svd(residuals, full_matrices=False)
        aw_periods, aw_series = residuals.shape
        aw_total_sum_squares = np.square(aw_singular_values).sum()
        aw_residual_variances = (
            aw_total_sum_squares
            - np.cumsum(np.square(aw_singular_values[:r_ic2]))
        ) / (aw_periods * aw_series)
        q_values = np.arange(1, r_ic2 + 1)
        aw_dimension_penalty = (aw_series + aw_periods) / (aw_series * aw_periods)
        aw_ic2 = np.log(aw_residual_variances) + q_values * aw_dimension_penalty * np.log(
            min(aw_series, aw_periods)
        )
        q_hat = int(np.argmin(aw_ic2) + 1)

        selection_rows.append(
            {
                "variant": variant,
                "sample": sample_name,
                "n_series": n_series,
                "n_months": data.shape[0],
                "r_ic1": r_ic1,
                "r_ic2": r_ic2,
                "r_ic3": r_ic3,
                "q_aw_ic2": q_hat,
            }
        )
        for r_value in r_values:
            bai_rows.append(
                {
                    "variant": variant,
                    "sample": sample_name,
                    "r": r_value,
                    "ic1": ic1[r_value - 1],
                    "ic2": ic2[r_value - 1],
                    "ic3": ic3[r_value - 1],
                }
            )
            bai_eigen_rows.append(
                {
                    "variant": variant,
                    "sample": sample_name,
                    "component": r_value,
                    "eigenvalue": np.square(singular_values[r_value - 1])
                    / (n_periods - 1),
                }
            )
        for q_value in q_values:
            aw_rows.append(
                {
                    "variant": variant,
                    "sample": sample_name,
                    "q": q_value,
                    "aw_ic2": aw_ic2[q_value - 1],
                }
            )
            aw_eigen_rows.append(
                {
                    "variant": variant,
                    "sample": sample_name,
                    "component": q_value,
                    "eigenvalue": np.square(aw_singular_values[q_value - 1])
                    / (aw_periods - 1),
                }
            )

python_selection = pd.DataFrame(selection_rows).sort_values(["variant", "sample"])
python_bai = pd.DataFrame(bai_rows).sort_values(["variant", "sample", "r"])
python_aw = pd.DataFrame(aw_rows).sort_values(["variant", "sample", "q"])
python_bai_eigen = pd.DataFrame(bai_eigen_rows).sort_values(
    ["variant", "sample", "component"]
)
python_aw_eigen = pd.DataFrame(aw_eigen_rows).sort_values(
    ["variant", "sample", "component"]
)

assert python_selection.reset_index(drop=True).equals(
    r_selection.sort_values(["variant", "sample"]).reset_index(drop=True)
)

comparisons = {
    "bai_ng_criteria": (
        python_bai[["ic1", "ic2", "ic3"]].to_numpy(),
        r_bai_ng.sort_values(["variant", "sample", "r"])[
            ["ic1", "ic2", "ic3"]
        ].to_numpy(),
    ),
    "aw_criteria": (
        python_aw[["aw_ic2"]].to_numpy(),
        r_aw.sort_values(["variant", "sample", "q"])[["aw_ic2"]].to_numpy(),
    ),
    "bai_ng_eigenvalues": (
        python_bai_eigen[["eigenvalue"]].to_numpy(),
        r_bai_eigen.sort_values(["variant", "sample", "component"])[
            ["eigenvalue"]
        ].to_numpy(),
    ),
    "aw_eigenvalues": (
        python_aw_eigen[["eigenvalue"]].to_numpy(),
        r_aw_eigen.sort_values(["variant", "sample", "component"])[
            ["eigenvalue"]
        ].to_numpy(),
    ),
}

comparison_rows = []
for quantity, (python_values, r_values) in comparisons.items():
    absolute_difference = np.abs(python_values - r_values)
    tolerance = 1e-10 + 1e-9 * np.maximum(np.abs(python_values), np.abs(r_values))
    comparison_rows.append(
        {
            "quantity": quantity,
            "n_values": python_values.size,
            "max_absolute_difference": absolute_difference.max(),
            "max_tolerance_ratio": np.max(absolute_difference / tolerance),
            "passed": bool(np.all(absolute_difference <= tolerance)),
        }
    )

comparison = pd.DataFrame(comparison_rows)
assert comparison["passed"].all()
comparison.to_csv(out_dir / "cross_language_comparison.csv", index=False)
python_selection.to_csv(out_dir / "python_selection.csv", index=False)
python_bai.to_csv(out_dir / "python_bai_ng_criteria.csv", index=False)
python_aw.to_csv(out_dir / "python_aw_criteria.csv", index=False)
print(comparison.to_string(index=False))

File: diagnostics/rq_dimension_audit/scripts/reproduce_factor_selection.R
source("R/data_download/panel_candidates.R")
source("R/preprocessing/panel_candidates.R")
source("R/modeling/factor_estimation.R")
source("R/modeling/impulse_response.R")
source("R/identification/experimental_panel.R")

out_dir <- "diagnostics/rq_dimension_audit/output"
panel_dir <- file.path(out_dir, "panels")
dir.create(panel_dir, recursive = TRUE, showWarnings = FALSE)

expected_dates <- seq(as.Date("2013-01-01"), as.Date("2025-09-01"), by = "month")
base <- readr::read_csv(
  "data/processed/data_log_deseasonalized_base_106.csv",
  show_col_types = FALSE
)
stopifnot(
  identical(as.Date(base$ref.date), expected_dates),
  nrow(base) == 153L,
  ncol(base) == 107L,
  !anyNA(base),
  all(is.finite(as.matrix(base[, -1])))
)

dates <- as.Date(base$ref.date)
base_mat <- base |>
  dplyr::select(-ref.date) |>
  as.matrix()
experimental <- build_factorial_drop_block_panels(
  base_mat,
  expected_dates,
  "yield_6m"
)
stopifnot(nrow(experimental$variant_manifest) == 64L)

samples <- list(
  full = dates <= as.Date("2025-09-01"),
  pre_covid = dates <= as.Date("2019-12-01")
)

selection_rows <- list()
bai_ng_rows <- list()
aw_rows <- list()
eigen_rows <- list()
aw_eigen_rows <- list()
sensitivity_rows <- list()
row_index <- 0L

for (variant in experimental$variant_manifest$variant) {
  panel <- experimental$panels[[variant]]$matrix
  readr::write_csv(
    tibble::as_tibble(panel) |>
      dplyr::mutate(ref.date = dates, .before = 1),
    file.path(panel_dir, paste0(variant, ".csv.gz"))
  )

  for (sample_name in names(samples)) {
    message("factor selection: ", variant, " / ", sample_name)
    data_sub <- panel[samples[[sample_name]], , drop = FALSE]
    diff_std <- diff(data_sub)
    diff_sd <- apply(diff_std, 2, sd)
    stopifnot(
      nrow(data_sub) %in% c(153L, 84L),
      all(is.finite(data_sub)),
      all(is.finite(diff_sd)),
      all(diff_sd > 0)
    )

    bai_ng <- bai_ng_criteria(data_sub, max_r = 20L, apply_bll = TRUE)
    r_ic2 <- bai_ng$r_hat$IC2
    aw <- amengual_watson(
      data_sub,
      r = r_ic2,
      p = 6L,
      max_q = r_ic2,
      apply_bll = TRUE
    )

    row_index <- row_index + 1L
    selection_rows[[row_index]] <- tibble::tibble(
      variant = variant,
      sample = sample_name,
      n_series = ncol(data_sub),
      n_months = nrow(data_sub),
      r_ic1 = bai_ng$r_hat$IC1,
      r_ic2 = r_ic2,
      r_ic3 = bai_ng$r_hat$IC3,
      q_aw_ic2 = aw$q_hat
    )
    bai_ng_rows[[row_index]] <- tibble::tibble(
      variant = variant,
      sample = sample_name,
      r = seq_len(20L),
      ic1 = bai_ng$criteria$IC1,
      ic2 = bai_ng$criteria$IC2,
      ic3 = bai_ng$criteria$IC3
    )
    eigen_rows[[row_index]] <- tibble::tibble(
      variant = variant,
      sample = sample_name,
      component = seq_len(20L),
      eigenvalue = bai_ng$pca$sdev[seq_len(20L)]^2
    )
    aw_rows[[row_index]] <- tibble::tibble(
      variant = variant,
      sample = sample_name,
      q = seq_len(r_ic2),
      aw_ic2 = aw$aw
    )

    x_std <- sweep(
      sweep(diff(data_sub), 2, colMeans(diff(data_sub)), "-"),
      2,
      apply(diff(data_sub), 2, sd),
      "/"
    )
    factor_scores <- prcomp(x_std, scale. = FALSE, center = FALSE)$x[, seq_len(r_ic2), drop = FALSE]
    lagged_factors <- stats::embed(factor_scores, 7L)
    design <- cbind(1, lagged_factors[, -seq_len(r_ic2), drop = FALSE])
    residuals <- x_std[7:nrow(x_std), , drop = FALSE] -
      design %*% solve(crossprod(design), crossprod(design, x_std[7:nrow(x_std), , drop = FALSE]))
    residual_pca <- prcomp(residuals, scale. = FALSE, center = FALSE)
    aw_eigen_rows[[row_index]] <- tibble::tibble(
      variant = variant,
      sample = sample_name,
      component = seq_len(r_ic2),
      eigenvalue = residual_pca$sdev[seq_len(r_ic2)]^2
    )

    conventions <- list(
      p6_cap_r = c(p = 6L, max_q = r_ic2),
      p6_max15 = c(p = 6L, max_q = 15L),
      p12_cap_r = c(p = 12L, max_q = r_ic2),
      p12_max15 = c(p = 12L, max_q = 15L)
    )
    sensitivity_rows[[row_index]] <- dplyr::bind_rows(lapply(
      names(conventions),
      function(convention) {
        setting <- conventions[[convention]]
        result <- tryCatch(
          amengual_watson(
            data_sub,
            r = r_ic2,
            p = setting[["p"]],
            max_q = setting[["max_q"]],
            apply_bll = TRUE
          ),
          error = function(error) error
        )
        if (inherits(result, "error")) {
          return(tibble::tibble(
            variant = variant,
            sample = sample_name,
            r_ic2 = r_ic2,
            convention = convention,
            p = setting[["p"]],
            max_q = setting[["max_q"]],
            q_hat = NA_integer_,
            q_gt_r = NA,
            failure = conditionMessage(result)
          ))
        }
        tibble::tibble(
          variant = variant,
          sample = sample_name,
          r_ic2 = r_ic2,
          convention = convention,
          p = setting[["p"]],
          max_q = setting[["max_q"]],
          q_hat = result$q_hat,
          q_gt_r = result$q_hat > r_ic2,
          failure = NA_character_
        )
      }
    ))
  }
}

selection <- dplyr::bind_rows(selection_rows) |>
  dplyr::arrange(variant, sample)
bai_ng_grid <- dplyr::bind_rows(bai_ng_rows) |>
  dplyr::arrange(variant, sample, r)
aw_grid <- dplyr::bind_rows(aw_rows) |>
  dplyr::arrange(variant, sample, q)
eigen_grid <- dplyr::bind_rows(eigen_rows) |>
  dplyr::arrange(variant, sample, component)
aw_eigen_grid <- dplyr::bind_rows(aw_eigen_rows) |>
  dplyr::arrange(variant, sample, component)
sensitivity <- dplyr::bind_rows(sensitivity_rows) |>
  dplyr::arrange(variant, sample, convention)

saved_selection <- readr::read_csv(
  "output/panel_experimental/rq_grid_drop_blocks/factor_selection/factor_selection_cells.csv",
  show_col_types = FALSE
) |>
  dplyr::select(-r_ic3_at_upper_bound, -p_aw)
saved_bai_ng <- readr::read_csv(
  "output/panel_experimental/rq_grid_drop_blocks/factor_selection/bai_ng_bll_criteria.csv",
  show_col_types = FALSE
)
saved_aw <- readr::read_csv(
  "output/panel_experimental/rq_grid_drop_blocks/factor_selection/amengual_watson_criteria.csv",
  show_col_types = FALSE
) |>
  dplyr::select(-r_ic2)

stopifnot(
  isTRUE(all.equal(selection, saved_selection, tolerance = 0, check.attributes = FALSE)),
  isTRUE(all.equal(bai_ng_grid, saved_bai_ng, tolerance = 1e-13, check.attributes = FALSE)),
  isTRUE(all.equal(aw_grid, saved_aw, tolerance = 1e-13, check.attributes = FALSE))
)

readr::write_csv(selection, file.path(out_dir, "r_selection.csv"))
readr::write_csv(bai_ng_grid, file.path(out_dir, "r_bai_ng_criteria.csv"))
readr::write_csv(aw_grid, file.path(out_dir, "r_aw_criteria.csv"))
readr::write_csv(eigen_grid, file.path(out_dir, "r_bai_ng_eigenvalues.csv"))
readr::write_csv(aw_eigen_grid, file.path(out_dir, "r_aw_eigenvalues.csv"))
readr::write_csv(sensitivity, file.path(out_dir, "aw_convention_sensitivity.csv"))

canonical_rows <- list()
test_rows <- list()
canonical_index <- 0L
test_index <- 0L
for (sample_name in names(samples)) {
  data_sub <- base_mat[samples[[sample_name]], , drop = FALSE]
  bai_ng <- bai_ng_criteria(data_sub, max_r = 20L, apply_bll = TRUE)
  r_ic2 <- bai_ng$r_hat$IC2
  for (p_lags in c(6L, 12L)) {
    for (max_q in c(r_ic2, 15L)) {
      aw <- amengual_watson(
        data_sub,
        r = r_ic2,
        p = p_lags,
        max_q = max_q,
        apply_bll = TRUE
      )
      canonical_index <- canonical_index + 1L
      canonical_rows[[canonical_index]] <- tibble::tibble(
        sample = sample_name,
        n_series = ncol(data_sub),
        n_months = nrow(data_sub),
        r_ic1 = bai_ng$r_hat$IC1,
        r_ic2 = r_ic2,
        r_ic3 = bai_ng$r_hat$IC3,
        p_aw = p_lags,
        max_q = max_q,
        q_aw_ic2 = aw$q_hat,
        q_gt_r = aw$q_hat > r_ic2
      )
    }
  }

  diff_std <- diff(data_sub)
  diff_std <- sweep(diff_std, 2, colMeans(diff_std), "-")
  diff_std <- sweep(diff_std, 2, apply(diff_std, 2, sd), "/")
  for (variable in colnames(diff_std)) {
    adf <- tseries::adf.test(diff_std[, variable], alternative = "stationary")
    pp <- tseries::pp.test(diff_std[, variable], alternative = "stationary")
    test_index <- test_index + 1L
    test_rows[[test_index]] <- tibble::tibble(
      sample = sample_name,
      variable = variable,
      adf_statistic = unname(adf$statistic),
      adf_p_value = adf$p.value,
      adf_reject_5pct = adf$p.value < 0.05,
      pp_statistic = unname(pp$statistic),
      pp_p_value = pp$p.value,
      pp_reject_5pct = pp$p.value < 0.05
    )
  }
}

canonical <- dplyr::bind_rows(canonical_rows)
stationarity <- dplyr::bind_rows(test_rows)
readr::write_csv(canonical, file.path(out_dir, "canonical_selection.csv"))
readr::write_csv(stationarity, file.path(out_dir, "canonical_difference_stationarity.csv"))

cat("R factor-selection audit passed.\n")
print(canonical)
print(
  sensitivity |>
    dplyr::group_by(sample, convention) |>
    dplyr::summarise(
      failures = sum(!is.na(failure)),
      q_gt_r = sum(q_gt_r, na.rm = TRUE),
      q_min = min(q_hat, na.rm = TRUE),
      q_max = max(q_hat, na.rm = TRUE),
      .groups = "drop"
    )
)
print(
  stationarity |>
    dplyr::group_by(sample) |>
    dplyr::summarise(
      n_series = dplyr::n(),
      adf_reject_5pct = sum(adf_reject_5pct),
      pp_reject_5pct = sum(pp_reject_5pct),
      both_reject_5pct = sum(adf_reject_5pct & pp_reject_5pct),
      neither_reject_5pct = sum(!adf_reject_5pct & !pp_reject_5pct),
      .groups = "drop"
    )
)
